client/common/zod.js

```diff
@@ -0,0 +1,29 @@
+import z from 'zod';
+
+/* eslint-disable */
+const EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/i;
+/* eslint-enable */
+
+const SignupFormInput = z.object({
+  username: z
+    .string()
+    .min(1)
+    .max(20)
+    .regex(/^[a-zA-Z0-9._-]{1,20}$/),
```

Comment:
The existing regex says that it must be between 1 and 20 characters and can only use certain characters: letters, numbers, dots, underscores and dashes.

The way that Zod breaks it up into different checks is very helpful because it allows us to show more specific messages, like “username cannot be longer than 20 characters” or “username contains invalid characters” (would be nice if we could say specifically which character).  But I don’t think we have translations for all of those, so this might be something where we make some improvements now but can continue to refine it over time.